packages/mui-lab/src/Masonry/Masonry.js

```diff
@@ -182,80 +182,95 @@ const Masonry = React.forwardRef(function Masonry(inProps, ref) {
-        }
-        if (!skip) {
-          // find the current shortest column (where the current item will be placed)
-          const currentMinColumnIndex = columnHeights.indexOf(Math.min(...columnHeights));
-          columnHeights[currentMinColumnIndex] += childHeight;
-          const order = currentMinColumnIndex + 1;
-          child.style.order = order;
-        }
-      });
-      if (!skip) {
-        setMaxColumnHeight(Math.max(...columnHeights));
-        const numOfLineBreaks = currentNumberOfColumns > 0 ? currentNumberOfColumns - 1 : 0;
-        setNumberOfLineBreaks(numOfLineBreaks);
-      }
-    };
 
+          const currentNumberOfColumns = Math.round(
+            parentWidth / (childWidth + firstChildMarginLeft + firstChildMarginRight),
+          );
+
+          const columnHeights = new Array(currentNumberOfColumns).fill(0);
+          let skip = false;
+          masonry.target.childNodes.forEach((child) => {
+            if (
+              child.nodeType !== Node.ELEMENT_NODE ||
+              child.dataset.class === 'line-break' ||
+              skip
+            ) {
+              return;
+            }
+            const childComputedStyle = window.getComputedStyle(child);
+            const childMarginTop = parseToNumber(childComputedStyle.marginTop);
+            const childMarginBottom = parseToNumber(childComputedStyle.marginBottom);
+            // if any one of children isn't rendered yet, masonry's height shouldn't be computed yet
+            const childHeight = parseToNumber(childComputedStyle.height)
+              ? Math.ceil(parseToNumber(childComputedStyle.height)) +
+                childMarginTop +
+                childMarginBottom
+              : 0;
+            if (childHeight === 0) {
+              skip = true;
+              return;
+            }
+            // if there is a nested image that isn't rendered yet, masonry's height shouldn't be computed yet
+            for (let i = 0; i < child.childNodes.length; i += 1) {
+              const nestedChild = child.childNodes[i];
+              if (nestedChild.tagName === 'IMG' && nestedChild.clientHeight === 0) {
+                observer.current.observe(nestedChild);
+                skip = true;
+                break;
+              }
+            }
+            if (!skip) {
+              // find the current shortest column (where the current item will be placed)
+              const currentMinColumnIndex = columnHeights.indexOf(Math.min(...columnHeights));
+              columnHeights[currentMinColumnIndex] += childHeight;
+              const order = currentMinColumnIndex + 1;
+              child.style.order = order;
+            }
+          });
+          if (!skip) {
+            setMaxColumnHeight(Math.max(...columnHeights));
+            const numOfLineBreaks = currentNumberOfColumns > 0 ? currentNumberOfColumns - 1 : 0;
+            setNumberOfLineBreaks(numOfLineBreaks);
+          }
+        }),
+  );
+
+  React.useEffect(() => {
```

Comment:
Maybe we should have just changed this to:
```suggestion
  React.useLayoutEffect(() => {
```

See https://github.com/mui-org/material-ui/issues/29209#issuecomment-958972556

Reply:
If you are suggesting that using `useLayoutEffect` makes it unnecessary to observe the first child, I just experimented and it still gives the [bug](https://github.com/mui-org/material-ui/pull/28059#discussion_r732537002).